File: problem_0139/Solution.java
/*
139. Word Break
    Total Accepted: 92410
    Total Submissions: 356258
    Difficulty: Medium

Given a string s and a dictionary of words dict, determine if s can be segmented
into a space-separated sequence of one or more dictionary words.

For example, given
s = "leetcode",
dict = ["leet", "code"].
Return true because "leetcode" can be segmented as "leet code". 
*/

import java.io.*;
import java.util.*;

public class Solution {
    public boolean wordBreak(String s, Set<String> wordDict) {
        int len = s.length();
        boolean[] res = new boolean[len];

        for (int i = 0; i < len; ++i)
        {
            if (wordDict.contains(s.substring(0,i+1)))
                res[i] = true;
            for (int j = 0; j < i && !res[i]; ++j)
            {
                if (res[j] && wordDict.contains(s.substring(j+1,i+1)))
                    res[i] = true;
            }
        }

        return res[len-1];
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        Set<String> dict;
        String s;

        dict = new HashSet<String>(Arrays.asList("leet","code"));
        s = "leetcode";
        System.out.println("Result: " + sol.wordBreak(s,dict));
    }
}

File: problem_0139/solution.cpp
/*
139. Word Break
Medium

Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, determine
if s can be segmented into a space-separated sequence of one or more dictionary words.

Note:

The same word in the dictionary may be reused multiple times in the segmentation.
You may assume the dictionary does not contain duplicate words.

Example 1:
Input: s = "leetcode", wordDict = ["leet", "code"]
Output: true
Explanation: Return true because "leetcode" can be segmented as "leet code".

Example 2:
Input: s = "applepenapple", wordDict = ["apple", "pen"]
Output: true
Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
             Note that you are allowed to reuse a dictionary word.
Example 3:
Input: s = "catsandog", wordDict = ["cats", "dog", "sand", "and", "cat"]
Output: false
*/

#include <algorithm>
#include <atomic>
#include <bitset>
#include <cassert>
#include <cmath>
#include <condition_variable>
#include <functional>
#include <future>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <list>
#include <map>
#include <memory>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <stack>
#include <string_view>
#include <thread>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        int sz = s.size();
        vector<bool> bs(sz, false);
        for (int i = 0; i < sz; ++i) {
            if (dict.count(s.substr(0, i+1))) {
                bs[i] = true;
            }

            for (int j = 0; j < i && !bs[i]; ++j) {
                bs[i] = bs[j] && dict.count(s.substr(j+1, i-j));
            }
        }
        return bs[sz-1];
    }
};

int main() {
    Solution sol;
    vector<string> wordDict;
    string s;

    // Output: true
    s = "leetcode", wordDict = {"leet", "code"};
    cout << boolalpha << sol.wordBreak(s, wordDict) << endl;

    // Output: true
    s = "applepenapple", wordDict = {"apple", "pen"};
    cout << boolalpha << sol.wordBreak(s, wordDict) << endl;

    // Output: false
    s = "catsandog", wordDict = {"cats", "dog", "sand", "and", "cat"};
    cout << boolalpha << sol.wordBreak(s, wordDict) << endl;

    return 0;
}
